### DISCUSSION TITLE
Remove the need for explicit calls to register(type:)

### DISCUSSION THREAD

**Discussion by pcbeard** [migueldeicaza/SwiftGodot#323] [Ideas]
I've come up with a way to discover the classes in a given executable image that need to be registered. My idea is to extend the `@Godot` macro like this:

```swift
/// All classes that use the `@Godot` macro will implicitly conform to this protocol, which acts as a runtime marker.
public protocol GodotMacroClass: GodotObject {}

/// Creates the definition for a Swift class to be surfaced to Godot.
///
/// This macro creates the required constructors that the SwiftGodot framework requires (the `init`, and the
/// `init(nativeHandle:)`) , ensures that both of those initialize the class if required, and registers
/// any `@Export` and `@Callable` methods for the class effectively surfacing properties and
/// methods to godot.
///
/// The macro also defines an extension to the class which conforms to `GodotMacroClass` defined above.
///
@attached(member,
          names: named (_initializeClass), named(classInitializer), named (implementedOverrides))
@attached(extension, conformances: GodotMacroClass)
public macro Godot() = #externalMacro(module: "SwiftGodotMacroLibrary", type: "GodotMacro")
```

In other words, every time the `@Godot` macro is used, the resulting class will also implicitly conform to a special protocol, `GodotMacroClass` (name TBD), which can be used at runtime to determine if a class type should be passed to `register(type:)`. This runtime introspection can be performed on Apple OSs using the runtime function `objc_enumerateClasses` like so:

```swift
func lookup_GodotMacroClasses() -> [GodotMacroClass.Type] {
    objc_enumerateClasses(subclassing: Wrapped.self).compactMap { c in
        c as? GodotMacroClass.Type
    }
}
```

I've devised a simple generic application framework that calls this automatically from the `runGodot()` callback:

```swift
    func registerTypes(level: GDExtension.InitializationLevel) {
        if level == .scene {
            let types = lookup_GodotMacroClasses()
            for type in types {
                register(type: type)
            }
        }
    }
```

I'm opening this topic for discussion, so that the people responsible for the other target platforms (Windows, Linux, etc.) can comment on this proposal. I'm open to alternative solutions that could work on platforms that have a much thinner runtime. Comments welcome.

**Comment by migueldeicaza** [comment]
Would love to get feedback from @tishin about how this might work on Windows.

**Comment by migueldeicaza** [comment]
There is a library that does some limited Swift reflection, that we might be able to extend to do this sort of lookups

https://github.com/wickwirew/Runtime

**Reply by pcbeard** [reply]
I've prototyped this (using static mirror reflection) on Linux using these APIs:

https://github.com/apple/swift/blob/main/include/swift-c/StaticMirror/
https://github.com/apple/swift/blob/main/include/swift-c/CommonString/

On an Ubuntu system (at least on Docker images I've tried), the shared library that implements this is here:

`/usr/lib/swift/linux/lib_InternalSwiftStaticMirror.so`

Does this same API exists in Swift on Windows?

**Reply by tishin** [reply]
@pcbeard I'm a bit confused. Are you referring to the `objc_enumerateClasses`-based example as the prototype here? Or is it some new prototype? I can check on physical Linux and Windows machines if you have a working package to test.

Though I was under impression that `objc_enumerateClasses` shouldn't be available outside Darwin.

**Reply by pcbeard** [reply]
On platforms where `objc_enumerateClasses()` is unavailable, we can emulate it using `swift_static_mirror_conformances_set_create()` to get a list of class names, and `NSClassFromString()` to look them up.

I envision a DEBUG mode where this happens at runtime, and using this information at build time to generate code that performs the registrations automatically. If the generated code is compiled in the same module that defines these classes, no dynamic lookup will be needed. `ModuleName.ClassName.self` can be used directly.

**Reply by pcbeard** [reply]
I checked with @compnerd who says he will be able to add the library so it can be used on a build machine.
